### PR TITLE
Fix #26: Implement application keys

### LINKED ISSUE #26: Add support for application keys APIs
The docs are live on the Backblaze web site, and the APIs will be available soon.  It's time for the SDK to support the new APIs.

### PR DESCRIPTION
**Pull request by bwbeach**

### CODE REVIEW

File: core/src/main/java/com/backblaze/b2/client/B2ListKeysIterable.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+package com.backblaze.b2.client;
+
+import com.backblaze.b2.client.structures.B2ApplicationKey;
+import com.backblaze.b2.client.structures.B2ListKeysRequest;
+
+import java.util.Iterator;
+
+/**
+ * This interface collects the APIs we provide on our B2FileVersion iterables.
```
Comment (certainmagic) [inline core/src/main/java/com/backblaze/b2/client/B2ListKeysIterable.java]: "B2FileVersion"?  :)

File: core/src/main/java/com/backblaze/b2/client/structures/B2ApplicationKey.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+package com.backblaze.b2.client.structures;
+
+import com.backblaze.b2.json.B2Json;
+
+import java.util.Objects;
+import java.util.TreeSet;
+
+/**
+ * Response from b2_create_key and b2_delete_key.  Included in response from b2_list_keys.
```
Comment (certainmagic) [inline core/src/main/java/com/backblaze/b2/client/structures/B2ApplicationKey.java]: "Response from b2_create_key".  i think that comment is a little out-of-date.  I think you're returning a B2CreatedApplicationKey from b2_create_key now.  :)

File: core/src/main/java/com/backblaze/b2/client/structures/B2Capability.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import com.backblaze.b2.json.B2JsonReader;
+import com.backblaze.b2.json.B2JsonTypeHandler;
+import com.backblaze.b2.json.B2JsonWriter;
+import com.backblaze.b2.util.B2StringUtil;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+public enum B2Capability {
```
Comment (certainmagic) [inline core/src/main/java/com/backblaze/b2/client/structures/B2Capability.java]: i'm a big fan of having strong typing when we can and enums are better for that than strings.  With that said...please read the "DESIGN NOTE" comment about compatibility in B2BucketTypes.java and then let's talk.  :)

File: core/src/main/java/com/backblaze/b2/client/structures/B2CreateKeyRequest.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+public class B2CreateKeyRequest {
+
+    private final Set<B2Capability> capabilies;
+    private final String keyName;
+    private final Long validDurationSeconds;
+    private final String bucketId;
+    private final String namePrefix;
+
+    private B2CreateKeyRequest(Set<B2Capability> capabilies,
```
Comment (certainmagic) [inline core/src/main/java/com/backblaze/b2/client/structures/B2CreateKeyRequest.java]: "capabilies"  :)

File: core/src/main/java/com/backblaze/b2/client/structures/B2CreatedApplicationKey.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+ */
+
+package com.backblaze.b2.client.structures;
+
+import com.backblaze.b2.json.B2Json;
+
+import java.util.Objects;
+import java.util.TreeSet;
+
+public class B2CreatedApplicationKey {
```
Comment (certainmagic) [inline core/src/main/java/com/backblaze/b2/client/structures/B2CreatedApplicationKey.java]: maybe comment that this is what's returned from b2_create_key and that it differs from the other version in that it has the secret (like your nice comment below for toApplicationKey()).

File: samples/src/main/java/com/backblaze/b2/sample/B2Sample.java [hunk 6]
```diff
@@ -73,6 +80,31 @@ private static void mainGuts(PrintWriter writer,
...
+        }
+
+        bigHeader(writer, "list application keys");
+        for (B2ApplicationKey key : client.applicationKeys()) {
+            writer.println("key id: " + key.getApplicationKeyId() + "   capabilities: " + key.getCapabilities());
+        }
+
+        bigHeader(writer, "delete application key");
+        {
+            client.deleteKey(B2DeleteKeyRequest.builder(applicationKeyId).build());
```
Comment (certainmagic) [inline samples/src/main/java/com/backblaze/b2/sample/B2Sample.java]: you're welcome to add a convenience "deleteKey(String appKeyId)" method to the storage client interface which would create the request and call the real method in the interface.

**Comment by bwbeach** [comment]
Thanks.  I agree about enum/string.  I'll change to string.

As for when to merge, I think Marj needs the updated SDK to build the UI.

- BrianB

**Comment by certainmagic** [comment]
We can use the new sdk without putting it in master.

**Comment by certainmagic** [comment]
I'm going to merge this to master even though the API isn't fully supported in the service yet.  I'll add a note to the README.

**Pull request merged**